### ISSUE TITLE
[12][XMLRPC] xrpc.client.Binary field not written on Odoo

### ISSUE THREAD

**Issue by luffah** [odoo/odoo#53637]
Impacted versions: 12

Steps to reproduce: 
write a binary field (a logo for example), with xmlrpclib (or oerplib) to odoo .
```
    ...
    with open('logo.png', 'r') as f:
        logodatas = f.read()
    logo = xmlrpclib.Binary(logodatas)

    oerp.execute('res.partner', 'create', {...., 'logo': logo})
```

Current behavior:
Crash with:
```
  File "<decorator-gen-3>", line 2, in create
  File "/home/odoo/odoo/parts/OCB-e5345c139f65369a3b3f2f7c317ffcc474276dc7/odoo/api.py", line 451, in _model_create_multi
    return create(self, [arg])
  File "/home/odoo/odoo/parts/OCB-e5345c139f65369a3b3f2f7c317ffcc474276dc7/odoo/models.py", line 3560, in create
    records = self._create(data_list)
  File "/home/odoo/odoo/parts/OCB-e5345c139f65369a3b3f2f7c317ffcc474276dc7/odoo/models.py", line 3646, in _create
    col_val = field.convert_to_column(val, self, stored)
  File "/home/odoo/odoo/parts/OCB-e5345c139f65369a3b3f2f7c317ffcc474276dc7/odoo/fields.py", line 1781, in convert_to_column
    if value[:1] == b'P':  # Fast detection of first 6 bits of '<' (0x3C)
TypeError: 'Binary' object is not subscriptable
```
The side effect is that external application that were compatible in 6 are no more compatible with Odoo 12 on this specific point.

Expected behavior:
Writing the logo as it could work with the following dirty patch.
```
▓█ diff odoo/fields.py{.bak,}
1780a1781,1787
>         # FIXME
>         import xmlrpc
>         if isinstance(value, xmlrpc.client.Binary):
>             import base64
>             value = base64.b64encode(value.data)
>         # /FIXME
> 
```

**Comment by pedrobaeza** [comment]
You have to encode yourself the binary content and pass it as regularly. This can be a wishlist, but there's no intention to be done IMO.

**Comment by luffah** [comment]
Sad for retro-compatibility.

**Comment by pedrobaeza** [comment]
@xmo-odoo do you think this should be handled at the first XML-RPC endpoint level?

**Comment by xmo-odoo** [comment]
@pedrobaeza I think we could just set `use_builtin_types` which was introduced in Python 3.3. It *might* even allow people to use "proper" datetimes over RPC as that's often what is used internally.

**Comment by pedrobaeza** [comment]
Yeah, good idea, should this be introduced in stable branches?

@luffah anyway, you would need to change your code for not using xmlrpclib.Binary

**Comment by xmo-odoo** [comment]
> Yeah, good idea, should this be introduced in stable branches?

I'll have to ask @odony but since we don't currently care for the Binary and Datetime wrappers the only thing it should do is... allow more stuff in and that's it.

> @luffah anyway, you would need to change your code for not using xmlrpclib.Binary

Not even! The flag is just used when deserialising: instead of using the xmlrpc wrappers it just deserialises to the native types, and that's it. There's no change in wire protocol.

**Comment by pedrobaeza** [comment]
Oh, cool, everyone wins! I wish all the issues can be solved in this satisfactory way.

**Comment by xmo-odoo** [comment]
> Oh, cool, everyone wins! I wish all the issues can be solved in this satisfactory way.

Turns out I spoke too soon: it only works with empty values, because the image processing pipeline still expects to get base64 data. Empty data is fine because an empty base64 and an empty not-base64 looks the same, but it chokes on non-empty data.

**Comment by pedrobaeza** [comment]
Well, any way the improvement would be welcome for those using XMLRPC intensively (I'm not one of them, as I avoid integrations. I prefer to spend budgets improving Odoo instead of connecting it with other software).

**Comment by luffah** [comment]
A future module that could be added in OCA  ?

Note : For my models, i use this function as a workaround.
```
import base64
try:
    from xmlrpc.client import Binary
except ImportError:
    #pylint: disable=bad-python3-import
    from xmlrpclib import Binary


def update_xmlrpc_values_on_binary_fields(vals, fields):
    """
       Update values to ensure mentionned binary fields can be stored in Odoo.

       Params:
          vals    - values as passed in write method
          fields  - name of binary fields
    """
    for f in fields:
        if vals.get(f, False) and isinstance(vals[f], Binary):
            vals[f] = base64.b64encode(vals[f].data)
    return vals
```

**Comment by luffah** [comment]
@pedrobaeza @xmo-odoo the issue was not the b64encoding : it is that Odoo can't manage xrpc.client.Binary field.

Sorry for the confusion.

The addition of b64encode was linked to wrong written tests cases. (which was using xmlrpclib through oerplib in python)

On the case i encounter issue, the binary is well encoded in base64 before sending. (using php-xmlrpc)
This case can be managed as described below.

**Comment by pedrobaeza** [comment]
@luffah `xrpc.client.Binary` objects mustn't be handled at that level, so the patch should be another.

**Comment by xmo-odoo** [comment]
> @pedrobaeza @xmo-odoo the issue was not the b64encoding : it is that Odoo can't manage xrpc.client.Binary field via xmlrpc write request.

It's really that you're trying to write binary data to an image field, and image fields are in base64. Supporting Binary doesn't change this specific issue, because you'd still have to encode to base64, which you can't do blindly because actual binary data is not *invalid*.

There are "raw" binary fields in Odoo but they are few and far between in Odoo, for various reasons historical and otherwise binary data is usually assumed to be base64. And the entire image-processing pipeline absolutely assumes that.

Just encode your data to base64 on the client side.

**Comment by luffah** [comment]
@xmo-odoo 
Yes, sorry again.... please consider my data is already encoded to base64 on the client side.
The wrong part (referencing base64) was commented in the issue.

My previous comment was for considered the issue at this point :
  - binary datas are encoded to base64 on the client side
  - a xrpc request is sent to write encoded datas as `xrpc.client.Binary` to odoo
  - odoo receive encapsulated (`xrpc.client.Binary') data but odoo reject it
 

@pedrobaeza 
on which level do you think it shall handled ? in `base/controllers/rpc.py` ?

**Comment by pedrobaeza** [comment]
Maybe, I don't know enough current RPC architecture

**Comment by luffah** [comment]
If it is managed in `odoo.addons.base.controllers.rpc`, then the modification to do could be really short : the fonction xmlrpc.client.loads shall be call with `use_builtin_types=True` introduced with python 3.3.

```python
# hijacking odoo method that parse params
from odoo.addons.base.controllers import rpc
from odoo.http import dispatch_rpc, request
from xmlrpc.client import dumps, loads

def _xmlrpc(self, service):
    """Common method to handle an XML-RPC request."""
    data = request.httprequest.get_data()
    params, method = loads(data, use_builtin_types=True)
    result = dispatch_rpc(service, method, params)
    return dumps((result,), methodresponse=1, allow_none=False)

rpc.RPC._xmlrpc = _xmlrpc
```

**Comment by xmo-odoo** [comment]
> If it is managed in `odoo.addons.base.controllers.rpc`, then the modification to do could be really short : the fonction xmlrpc.client.loads shall be call with `use_builtin_types=True` introduced with python 3.3.

Yes, but if you're properly sending base64 data from the client you have to b64encode it, so why not decode it to a string also. This might also be the source of the issue: with a python 2 client, encoding to base64 would yield `str` which is bytes but would be sent as string in the xmlrpc. With a python 3 client base64 encoding would yield `bytes` which would be sent as binary over the wire (and received as a xmlrpc Binary).

We can set `use_builtin_types` and that's unlikely to do harm either way, but I'm a bit concerned that you'll still have issues down the line if you try interacting with an odoo running on python 2 from a client running on python 3.
